
# EvoGen_course

Evolutionary Genomics course - Data Analysis module

13th-17th April 2015

## Material

Instructions to download all material necessary for the practical sessions are given [here](https://github.com/mfumagalli/EvoGen_course/blob/master/install.md).

## Agenda

### Day 1

 * Lecture: From raw data to SNPs ([pdf](https://github.com/mfumagalli/EvoGen_course/blob/master/slides_day_1.pdf)).

 * Journal discussion:

 *	Practical session:
 
	[data filtering](https://github.com/mfumagalli/EvoGen_course/blob/master/filtering.md)

	[SNP calling](https://github.com/mfumagalli/EvoGen_course/blob/master/snpcall.md)

 * Additional material: SNP calling using [SAMtools](https://github.com/mfumagalli/EvoGen_course/blob/master/snpcall_samtools.md).

### Day 2

 - Lecture: Genotype calling and advanced methods for low-quality data ([pdf](https://github.com/mfumagalli/EvoGen_course/blob/master/slides_day_2.pdf)).

 - Journal discussion:

 - Practical session: [genotype calling](https://github.com/mfumagalli/EvoGen_course/blob/master/genocall.md), advanced methods to estimate [SFS](https://github.com/mfumagalli/EvoGen_course/blob/master/sfs.md)

 - Additional material: genotype calling using ([SAMtools](https://github.com/mfumagalli/EvoGen_course/blob/master/genocall_samtools.md)) and ([BEAGLE](https://github.com/mfumagalli/EvoGen_course/blob/master/imputation.md)), estimation of [inbreeding](https://github.com/mfumagalli/EvoGen_course/blob/master/inbreeding.md) coefficients, advanced methods to calculate summary statistics using [ANGSD](https://github.com/mfumagalli/EvoGen_course/blob/master/lowcov.md), [ngsPopGen](https://github.com/mfumagalli/EvoGen_course/blob/master/lowcov_ngstools.md) and [ngsTools](https://github.com/mfumagalli/ngsTools/blob/master/TUTORIAL.md)

Suggested reading before the workshop is Nielsen et al. Nature Reviews Genetics 2011, available [here](http://cteg.berkeley.edu/~nielsen/wordpress/wp-content/uploads/2013/01/Nielsen-R.-et-al.-2011.pdf).

### Day 3

 - Lecture: Population structure and effect of demography on genetic diversity

 - Journal discussion:

 - Practical exercises: PCA/MDS, ABC/dadi

### Day 4

 - Lecture: Natural selection

 - Journal discussion:

 - Practical exercises: scan FST/PBS, nSL/iHS H1...


## Credits

Some materials have been borrowed (and then adapted) from Thorfinn Korneliussen, Tyler Linderoth, Filipe G. Vieira.
Sequencing data on butterfiles has been kindly provided by Zach Gompert.<br>
ANGSD has been developed by Thorfinn Korneliussen and Anders Albrechtsen. 
Filipe G. Vieira implemented the inbreeding estimation. 
Tyler Linderoth wrote most of data filtering scripts together with Sonal Singhai, Ke Bi, and Filipe G. Vieira.








